src/vos/tests/evt_ctl.c

```diff
@@ -1223,6 +1223,134 @@ test_evt_iter_delete(void **state)
+	rc = evt_create(EVT_FEAT_DEFAULT, 13, arg->ta_uma, arg->ta_root,
+			DAOS_HDL_INVAL, &toh);
+	assert_int_equal(rc, 0);
+	srand(time(0));
+	hole_epoch = (rand() % 98) + 1;
+	print_message("Hole inserted %d epoch.\n", hole_epoch);
+	/* Insert data : deadbeef */
+	for (epoch = 1; epoch <= NUM_EPOCHS; epoch++) {
+		for (offset = epoch;
+			offset < sizeof(testdata) + epoch;
+			offset += sizeof(testdata)) {
+			entry.ei_rect.rc_ex.ex_lo = offset;
+			entry.ei_rect.rc_ex.ex_hi = offset+sizeof(testdata);
+			entry.ei_rect.rc_epc = epoch;
+			memset(&entry.ei_csum, 0, sizeof(entry.ei_csum));
+			entry.ei_ver = 0;
+			/* Insert a hole at random epoch */
+			if (epoch == hole_epoch)
+				entry.ei_inob = 0;
+			else
+				entry.ei_inob = sizeof(testdata);
+			if (entry.ei_inob == 0) {
+				rc = bio_alloc_init(arg->ta_utx,
+						&entry.ei_addr, NULL, 0);
+				assert_int_equal(rc, 0);
+			} else {
+				rc = bio_alloc_init(arg->ta_utx,
+						&entry.ei_addr,
+						testdata, sizeof(testdata));
+				assert_int_equal(rc, 0);
+			}
+			rc = evt_insert(toh, &entry);
+			assert_int_equal(rc, 0);
+		}
+	}
+	/*Prepare and Probe the tree. Iteration flags not set */
+	rc = evt_iter_prepare(toh, 0, NULL, &ih);
+	assert_int_equal(rc, 0);
+	rc = evt_iter_probe(ih, EVT_ITER_FIRST, NULL, NULL);
+	assert_int_equal(rc, 0);
+	/*
+	 * Delete each record from last and run evt_find
+	 * you get deadbeef, d (2-records). Covered records
+	 * should be exposed on each deletes
+	 */
+	for (epoch = NUM_EPOCHS; epoch > 0; epoch--) {
+		rect.rc_ex.ex_lo = epoch-1;
+		rect.rc_ex.ex_hi = epoch+9;
+		rect.rc_epc = epoch;
+		evt_ent_array_init(&ent_array);
+		rc = evt_find(toh, &rect, &ent_array);
+		if (rc != 0)
+			D_FATAL("Add rect failed %d\n", rc);
+		evt_ent_array_for_each(ent, &ent_array) {
+			bool punched;
+			static char buf[10];
+
+			addr = ent->en_addr;
+			strcpy(buf, " ");
+			punched = bio_addr_is_hole(&addr);
+			D_PRINT("Find rect "DF_ENT" width=%d val=%.*s\n",
+			DP_ENT(ent),
+			(int)evt_extent_width(&ent->en_sel_ext),
+			punched ? 4 : (int)evt_extent_width(&ent->en_sel_ext),
+			punched ? "None" : (char *)utest_off2ptr(arg->ta_utx,
+								 addr.ba_off));
+			punched ? strcpy(buf, "None") :
+			strncpy(buf,
+				(char *)utest_off2ptr(arg->ta_utx, addr.ba_off),
+				(int)evt_extent_width(&ent->en_sel_ext));
+			if (punched) {
+				rc = strcmp(buf, "None");
+				if (rc != 0)
+					fail_msg("Data Check Failed");
+			} else {
+				if ((int)evt_extent_width(&ent->en_sel_ext)
+					== 1) {
+					rc = strcmp(buf, "d");
+					if (rc != 0)
+						fail_msg("Data Check Failed");
+				} else {
+					rc = strcmp(buf, "deadbeef");
+					if (rc != 0)
+						fail_msg("Data Check Failed");
+				}
+			}
+		}
+		/* Delete the last visible record */
+		entry.ei_rect.rc_ex.ex_lo = epoch;
+		entry.ei_rect.rc_ex.ex_hi = rect.rc_ex.ex_hi;
+		entry.ei_rect.rc_epc = rect.rc_epc;
+		rc = evt_delete(toh, &entry.ei_rect, NULL);
+		assert_int_equal(rc, 0);
+	}
+	evt_ent_array_fini(&ent_array);
```

Comment:
This should be in the same scope as evt_ent_array_init.   Otherwise, it's a leak.

Reply:
Good finds... Thanks... Done.